pkg/core/klinedriver/klinebuilder.go

```diff
@@ -0,0 +1,195 @@
+// The rationale is that we want to keep the logic as simple as possible and let the driver/caller
+// to handle subtle adjustments, such as adjusting the end time to be 1ms before the next start time.
+// Also, it will make it easier to write test for the KLineBuilder.
+func (kb *KLineBuilder) Update(updateTime types.Time) (kLinesMap map[types.Interval][]*types.KLine) {
+	kLinesMap = make(map[types.Interval][]*types.KLine)
+	for interval, bufferedKLines := range kb.klinesBuffer {
+		for _, klineRef := range bufferedKLines {
+			kline := *klineRef
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+		}
+		kb.klinesBuffer[interval] = make([]*types.KLine, 0)
+	}
+	for interval, accBucket := range kb.accBucketMap {
+		if updateTime.Before(accBucket.StartTime.Time()) {
+			// update time is before the start time, just add kline
+			kline := *accBucket.KLine
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+			continue
+		}
+
+		// check if kline is available: there were trades or we have the last kline
+		lastBucket, hasLast := kb.lastBucketMap[interval]
+		if accBucket.KLine.NumberOfTrades == 0 && !hasLast {
+			log.Debugf("kline not available for interval %s(%s) yet", kb.symbol, interval)
+			continue
+		}
+
+		// At this point, either there were trades or we have the last kline
+		if accBucket.KLine.NumberOfTrades == 0 {
+			// no trade happened up to the updateTime
+			// use the last kline to fill the gap
+			accBucket.KLine.Open = lastBucket.KLine.Close
+			accBucket.KLine.High = lastBucket.KLine.Close
+			accBucket.KLine.Low = lastBucket.KLine.Close
+			accBucket.KLine.Close = lastBucket.KLine.Close
+			accBucket.KLine.LastTradeID = lastBucket.KLine.LastTradeID
+			accBucket.KLine.Volume = fixedpoint.Zero
+			accBucket.KLine.QuoteVolume = fixedpoint.Zero
+			accBucket.KLine.TakerBuyBaseAssetVolume = fixedpoint.Zero
+			accBucket.KLine.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+		}
+		// two cases from here:
+		// - the update time is in the accumulating bucket
+		// - the update time is after the accumulating bucket
+		//   - there might be multiple buckets in between
+
+		// check if the update time is in the accumulating bucket
+		if accBucket.Contains(updateTime) {
+			// update time is in the accumulating bucket, just add kline
+			accBucket.KLine.EndTime = types.Time(updateTime.Time().UTC())
+			kline := *accBucket.KLine
+			kLinesMap[interval] = append(kLinesMap[interval], &kline)
+			continue
+		}
+
+		// update time is after the accumulating bucket
+		// we should close the current accumulating bucket and add kline
+		accBucket.KLine.Closed = true
+		accBucket.KLine.EndTime = types.Time(accBucket.EndTime.Time().UTC())
+		kline := *accBucket.KLine
+		kLinesMap[interval] = append(kLinesMap[interval], &kline)
+		accBucket.Exposed = true
+		// find the next containing bucket and reset the accumulating bucket
+		nextBucket, numShifts := accBucket.findNextBucket(updateTime)
```

Comment:
Just a small suggestion:
since we already iterate through the gaps in findNextBucket, why not return an array of buckets within the gap?
The FillGapKLines function then could be removed.

Reply:
Leave the comment open as a reminder of TODO items

Reply:
ok

Reply:
Done.
I leave a comment as doc string of `Bucket.findNextBucket()`.